frontends/concrete-python/examples/sha256/main.py

```diff
@@ -0,0 +1,413 @@
+
+        carry = result_with_overflow >> CHUNK_WIDTH
+        result = result_with_overflow - (carry * (2**CHUNK_WIDTH))
+
+        results.append(result)
+    results = list(reversed(results))
+
+    if all(isinstance(x, (int, np.integer)) for x in results):
+        return np.array(results)
+
+    return fhe.array(results)
+
+
+def right_rotate_chunks(chunks, amount):
+    return np.concatenate((chunks[-amount:], chunks[:-amount]))
+
+
+def right_shift_chunks(chunks, amount):
+    return np.concatenate(([0] * chunks[-amount:].shape[0], chunks[:-amount]))
+
+
+def right_rotate_number(chunks, amount):
+    n = amount // CHUNK_WIDTH
+    m = amount % CHUNK_WIDTH
+
+    if n != 0:
+        rotated_chunks = right_rotate_chunks(chunks, amount=n)
+    else:
+        rotated_chunks = chunks
+
+    if m != 0:
+        raise_low_bits = lambda x: (x % 2**m) << (CHUNK_WIDTH - m)
+        raised_low_bits = fhe.univariate(raise_low_bits)(rotated_chunks)
+        shifted_raised_low_bits = right_rotate_chunks(raised_low_bits, amount=1)
+        high_bits = rotated_chunks >> m
+        rotated_number = shifted_raised_low_bits + high_bits
+    else:
+        rotated_number = rotated_chunks
+
+    return rotated_number
+
+
+def right_shift_number(chunks, amount):
+    n = amount // CHUNK_WIDTH
+    m = amount % CHUNK_WIDTH
+
+    if n != 0:
+        shifted_chunks = right_shift_chunks(chunks, n)
+    else:
+        shifted_chunks = chunks
+
+    if m != 0:
+        raise_low_bits = lambda x: (x % 2**m) << (CHUNK_WIDTH - m)
+        raised_low_bits = fhe.univariate(raise_low_bits)(shifted_chunks)
+        shifted_raised_low_bits = right_shift_chunks(raised_low_bits, 1)
+        high_bits = shifted_chunks >> m
+        result = shifted_raised_low_bits + high_bits
+    else:
+        result = shifted_chunks
+
+    return result
+
+
+def s0(w):
+    return right_rotate_number(w, 7) ^ right_rotate_number(w, 18) ^ right_shift_number(w, 3)
+
+
+def s1(w):
+    return right_rotate_number(w, 17) ^ right_rotate_number(w, 19) ^ right_shift_number(w, 10)
+
+
+def S0(a_word):
+    return (
+        right_rotate_number(a_word, 2)
+        ^ right_rotate_number(a_word, 13)
+        ^ right_rotate_number(a_word, 22)
+    )
+
+
+def S1(e_word):
+    return (
+        right_rotate_number(e_word, 6)
+        ^ right_rotate_number(e_word, 11)
+        ^ right_rotate_number(e_word, 25)
+    )
+
+
+def Ch(e_word, f_word, g_word):
+    return (e_word & f_word) ^ ((2**CHUNK_WIDTH - 1 - e_word) & g_word)
+
+
+def Maj(a_word, b_word, c_word):
+    return (a_word & b_word) ^ (a_word & c_word) ^ (b_word & c_word)
+
+
+def loop(state, w_i_plus_k_i):
+    a, b, c, d, e, f, g, h = state
+
+    temp1 = add(h, S1(e), Ch(e, f, g), w_i_plus_k_i)
+    temp2 = add(S0(a), Maj(a, b, c))
+
+    new_a = add(temp1, temp2)
+    new_e = add(d, temp1)
+
+    return [new_a, a, b, c, new_e, e, f, g]
+
+
+def sha256(data, number_of_rounds=64):
+    h_chunks = deepcopy(H_CHUNKED)
```

Comment:
np.array instead of copy ?
h_chunks is modified so why not but why copying k_chunks ?

Reply:
Nope, `np.array` has a single dtype, we want tracers of vectors in some rows and np.ndarray in other rows:
```
[
    np.array([1, 2, 3]),
    np.array([4, 5, 6]),
    Tracer,
    Tracer,
    np.array([x, y, z]),
]
```

Reply:
> h_chunks is modified so why not but why copying k_chunks ?
Not necessary but looks better :stuck_out_tongue: